### PR TITLE
project: fix build rpm

### PR DESCRIPTION
**Pull request by LK4D4**
- change workdir for accessing install-binaries.sh
- use other gopath for binaries to preserve sources
- add sources of proxy and grimes to rpc spec
- use dynamic proxy with -linkmode external in deb and rpm

Closes #27858

### CODE REVIEW

**Comment by clnperez** [comment]
@LK4D4 PTAL at PR #27858 and the comments there.

**Comment by LK4D4** [comment]
ping @clnperez @tiborvass @justincormack @andrewhsu

File: hack/dockerfile/install-binaries.sh [hunk 1]
```diff
@@ -32,6 +41,14 @@ install_containerd() {
 	cp bin/ctr /usr/local/bin/docker-containerd-ctr
 }
 
+install_proxy() {
+	echo "Install docker-proxy version $LIBNETWORK_COMMIT"
+	git clone https://github.com/docker/libnetwork.git "$GOPATH/src/github.com/docker/libnetwork"
+	cd "$GOPATH/src/github.com/docker/libnetwork"
+	git checkout -q "$LIBNETWORK_COMMIT"
+	CGO_ENABLED=$CGO_ENABLED go build -ldflags="$PROXY_LDFALGS" -o /usr/local/bin/docker-proxy github.com/docker/libnetwork/cmd/proxy
```
Comment (tiborvass) [inline hack/dockerfile/install-binaries.sh]: Isn't `CGO_ENABLED=$CGO_ENABLED` useless? :)

Comment (LK4D4) [inline hack/dockerfile/install-binaries.sh]: I have no idea how that bash stuff works :D I thought you export it or use it inline, just variables not used by spawned command, I might be wrong, though.

File: hack/dockerfile/install-binaries.sh [hunk 2]
```diff
@@ -68,11 +85,11 @@ do
...
-			echo "Install docker-proxy version $LIBNETWORK_COMMIT"
-			git clone https://github.com/docker/libnetwork.git "$GOPATH/src/github.com/docker/libnetwork"
-			cd "$GOPATH/src/github.com/docker/libnetwork"
-			git checkout -q "$LIBNETWORK_COMMIT"
-			CGO_ENABLED=0 go build -v -o /usr/local/bin/docker-proxy github.com/docker/libnetwork/cmd/proxy
+			CGO_ENABLED=0 install_proxy
+			;;
+
+		proxy-dynamic)
+			CGO_ENABLED=1 PROXY_LDFLAGS="-linkmode external" install_proxy
```
Comment (tiborvass) [inline hack/dockerfile/install-binaries.sh]: An equal sign is missing, it should be `-linkmode=external`. Might not be super important but it's safer, also it's how it was previously.

And do you confirm that it doesn't work with cgo disabled `CGO_ENABLED=0 PROXY_LDFLAGS='-linkmode=external'` ?

Comment (LK4D4) [inline hack/dockerfile/install-binaries.sh]: It's probably works, but this supposed to be dynamic binary, so why bother about CGO?

Comment (LK4D4) [inline hack/dockerfile/install-binaries.sh]: btw, it's used without equal sign everywhere else
I'll change.

Comment (tiborvass) [inline hack/dockerfile/install-binaries.sh]: Was just curious :) Is the linkmode=external automatically enabling cgo? Also this binary itself should not have any cgo in it.

**Comment by clnperez** [comment]
I'm still getting a build ID note error :crying_cat_face: 
```
extracting debug info from /root/rpmbuild/BUILDROOT/docker-engine-1.13.0-0.0.20161027.235256.git55249e2.fc23.x86_64/usr/bin/docker-proxy
*** ERROR: No build ID note found in /root/rpmbuild/BUILDROOT/docker-engine-1.13.0-0.0.20161027.235256.git55249e2.fc23.x86_64/usr/bin/docker-proxy
error: Bad exit status from /var/tmp/rpm-tmp.2Xw8DC (%install)
```

**Comment by tiborvass** [comment]
Looks good, but @andrewhsu if you could test it please :)

**Comment by andrewhsu** [comment]
Yup, I'll have a look.

**Comment by clnperez** [comment]
Looks like the latest update took care of that. Looks good (f23 rpms built)!

**Comment by LK4D4** [comment]
@clnperez nice

**Comment by andrewhsu** [comment]
LGTM

I've verified building of fedora-24 rpms produces stuff:
```
$ time make DOCKER_BUILD_PKGS=fedora-24 rpm
...
real    28m56.334s
user    0m2.738s
sys     0m1.091s
$ docker run --rm -i -t \
  -v `pwd`/bundles/latest/build-rpm/fedora-24/RPMS/x86_64:/r \
  --entrypoint=/bin/rpm fedora:24 -qpl /r/docker-engine-*.rpm \
  | grep -i '^/usr/bin'
/usr/bin/docker
/usr/bin/docker-containerd
/usr/bin/docker-containerd-ctr
/usr/bin/docker-containerd-shim
/usr/bin/docker-init
/usr/bin/docker-proxy
/usr/bin/docker-runc
/usr/bin/dockerd
```

**Comment by tiborvass** [comment]
LGTM

**Review by thaJeztah — APPROVED** [review]
I see that both @clnperez and @andrewhsu verified that this is working correctly, so I'll LGTM

thanks both for reviewing!

**Pull request merged**